### DISCUSSION TITLE
Erro ao abrir arquivo json no flutter

### DISCUSSION THREAD

**Discussion by ghost** [Flutterando/forum#96] [Perguntas e Respostas]
Galera estou iniciando no flutter e estou aprendendo pelos videos do youtube e algumas pesquisas na internet
porém cheguei ao Mapeamento de objetos Json consegui listar os objetos json em uma string porem ao jogar em um arquivo chamado dados.json ele retorna o seguinte erro abaixo:

/flutter ( 7193): https://github.com/Flutterando/forum/discussions/8 main package:mapeando_arquivo_json02/main.dart:6 E/flutter ( 7193): https://github.com/Flutterando/forum/discussions/9 _runMain. (dart:ui/hooks.dart:134:23) E/flutter ( 7193): #10 _delayEntrypointInvocation. (dart:isolate-patch/isolate_patch.dart:297:19) E/flutter ( 7193): https://github.com/Flutterando/forum/discussions/11 _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:192:12) E/flutter ( 7193):

meu projeto é este no pastebin.
https://pastebin.com/vqqngci3

alguem poderia me ajudar a encontrar o problema e abrir o arquivo listando todos os dados ja fazem dias que nao estou conseguindo.

**Comment by EuHigorBarbosa** [comment]
Então...eu implementei aqui no projeto que estou desenvolvendo e deu certo. 
A diferença é que tive que usar 

String response = await rootBundle.rootBundle.loadString('lib/assets/json/json.json');

Tive que usar o rootBundle.rootBundle pois só assim para ter acesso ao loadString:
![image](https://user-images.githubusercontent.com/87345837/206031793-ed2ae0f4-9925-4fb3-934b-eb5d1789e327.png)


E minha pasta de assets está na lib ...olha a foto da organização da pasta e do pubSpec
![image](https://user-images.githubusercontent.com/87345837/206031393-49a35bcc-5093-4385-b0ec-48dcadeef011.png)

Abaixo o resultado do print:
![image](https://user-images.githubusercontent.com/87345837/206031512-6f130447-4be7-4ad6-8743-b12c3af4353f.png)

Veja se ajudou...

**Comment by ghost** [comment]
Aqui ainda não deu certo o estranho é que na função:

 Future<Usuario> loadUsers() async {
  String jsonString = await loadAsset();
  final jsonResponse = json.decode(jsonString);
 
  //Aqui ele só retorna o primeiro usuario caso o arquivo seja aberto
  return Usuario.fromJson(jsonResponse[0]);
}

foi baseada em um outro projeto que eu vi ele tambem abre um arquivo que 
tem a mesma estrutura deste e mesmo local assets/json/arquivo.json,
porem só tem dois usuarios ele esta abrindo o arquivo e como esta especificando zero retorna o primeiro usuario se eu quiser o segundo tenho que mudar dentro da função o jsonResponse[1], mas retornando apenas um usuario o projeto é um pouco mais complexo,
eu realmente nao entendi o por que nao esta dando certo de abrir o arquivo e listar todos os usuarios.

**Reply by ewertonls** [reply]
Ele está retornando só um porque você está selecionando um item específico da lista.

Faz assim que deve funcionar:

```dart
return List.from(jsonResponse).map(
  (e) => User.fromJson(e),
).toList()
```

E também coloca a tipagem da função: `Future<List<User>>`

**Reply by EuHigorBarbosa** [reply]
Resolveu?

**Reply by ghost** [reply]
Não deu certo por que o outro projeto nao tem nada a ver com esse coloquei a função ai só pra exemplificar e fazer  com base nesta função dentro da Main esta função nao esta sendo usada a forma que eu estou tentando fazer é esta abaixo:

void main() async{

String jsonData = await loadAsset();

// Transforma o JSON numa lista de Maps.
final lista = jsonDecode(jsonData) as List;

// Transformando a lista de Maps em uma lista do tipo da sua classe Usuario
final List listaDeTodos = lista.map((item) => Usuario.fromJson(item)).toList();

// Para passo item a item
for (final element in listaDeTodos) {
print("Título: " + element.title);
}

}

mas esta retornando o seguinte erro abaixo:

E/flutter ( 7053): #10 _runMain. (dart:ui/hooks.dart:134:23)
E/flutter ( 7053): https://github.com/Flutterando/forum/discussions/11 _delayEntrypointInvocation. (dart:isolate-patch/isolate_patch.dart:297:19)
E/flutter ( 7053): https://github.com/Flutterando/forum/discussions/12 _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:192:12)

**Reply by ewertonls** [reply]
@wikianon está difícil de entender o que está acontecendo aí, fiz um projeto dart e funcionou normalmente. O arquivo `dados.json` está na raiz do projeto. 
```dart
import 'dart:io';
import 'dart:convert';

void main() async {
  final users = await loadUsers();
  print(users);
}

Future<List<Usuario>> loadUsers() async {
  final jsonString = await loadFile();
  final map = jsonDecode(jsonString);
  final users = List.from(map).map((e) => Usuario.fromJson(e)).toList();
  return users;
}

Future<dynamic> loadFile() async {
  return await File('dados.json').readAsString();
}

class Usuario {
  final int userId;
  final int id;
  final String title;
  final bool completed;

  const Usuario({
    required this.userId,
    required this.id,
    required this.title,
    required this.completed,
  });

  static Usuario fromJson(Map<String, dynamic> json) {
    return Usuario(
      userId: json['userId'],
      id: json['id'],
      title: json['title'],
      completed: json['completed'],
    );
  }
}
```

**Reply by ghost** [reply]
eu não sei o que esta acontecendo ainda nao deu certo veja como esta no meu projeto: 
![pubspec](https://user-images.githubusercontent.com/25127502/206591807-fe19bf0f-9f96-41fe-a5b8-33bb24423e55.png)

a minha main esta do mesmo jeito da sua public
![main](https://user-images.githubusercontent.com/25127502/206591831-0324236b-64ab-4918-89c2-048c808389b8.png)

meu assets:

![assets](https://user-images.githubusercontent.com/25127502/206593338-602feffa-63ae-46dc-9a4e-dcc5ed59405f.png)

dart sdk:

![dart_sdk](https://user-images.githubusercontent.com/25127502/206593399-ced1a28b-0a80-49b1-86d9-9f8efe93445d.png)

**Comment by EuHigorBarbosa** [comment]
Que tal se vc fizesse o seguinte....pra facilitar pra quem deseja te ajudar....sobe o projeto q está dando erro pro github. E aí a gente retira o bug dele e te conta o q fizemos....q tal?

Pois eu fiz aqui no meu e deu certo...consegui acessar o dado do arquivo json.

**Reply by ghost** [reply]
Hoje consegui upar o projeto esses dias atras nao estava em casa aqui esta o link do projeto.

https://github.com/wikianon/example_aluno_file_json
